ui/lib/utils/tableColumnFactory.tsx

```diff
@@ -0,0 +1,256 @@
+import { Tooltip } from 'antd'
+import { max as _max } from 'lodash'
+import {
+  IColumn,
+  ColumnActionsMode,
+} from 'office-ui-fabric-react/lib/DetailsList'
+import React from 'react'
+import { getValueFormat } from '@baurine/grafana-value-formats'
+
+import {
+  Bar,
+  Pre,
+  TextWithInfo,
+  TextWrap,
+  DateTime,
+  HighlightSQL,
+  IColumnKeys,
+} from '@lib/components'
+
+type Bar = { [key: string]: string }
+type BarsConfig = {
+  displayTransKey?: string // it is same as avg field name default
+  bars: [Bar, Bar, Bar?] // [avg, max, min?]
+}
+
+export type IExtendColumn = IColumn & {
+  refFields?: string[]
+}
+
+function capitalize(s: string) {
+  return s.charAt(0).toUpperCase() + s.slice(1)
+}
+
+export function formatVal(val: number, unit: string) {
+  const formatFn = getValueFormat(unit)
+  return unit === 'short' ? formatFn(val, 0, 1) : formatFn(val, 1)
+}
+
+export function commonColumnName(transPrefix: string, fieldName: string): any {
+  const fullTransKey = `${transPrefix}.${fieldName}`
+  return <TextWithInfo.TransKey transKey={fullTransKey} />
+}
+
+export class TableColumnFactory {
+  transPrefix: string
+  bar: BarColumn
+
+  constructor(transKeyPrefix: string) {
+    this.transPrefix = transKeyPrefix
+    this.bar = new BarColumn(this)
+  }
+
+  columnName(fieldName: string): any {
+    return commonColumnName(this.transPrefix, fieldName)
+  }
+
+  textWithTooltip(fieldName: string): IExtendColumn {
```

Comment:
Maybe we can preserve the type check:

```typescript
textWithTooltip<T extends string>(fieldName: T, _rows: { [K in T]?: any }[] ) ...
```

Test:

```typescript
function textWithTooltip<T extends string>(_fieldName: T, _rows: { [K in T]?: any }[] ) {

}

type X = {
    def: '123'
}

function testX(a: X) {
    textWithTooltip("abc", [a]) // Error: Type 'X' has no properties in common with type '{ abc?: any; }'.
}

type Y = {
    abc?: '123'
}

function testY(a: Y) {
    textWithTooltip("abc", [a])
}
```

Reply:
It would be interesting to see how this type check will be implemented for multi bars :)

https://stackoverflow.com/questions/49285864/is-there-a-valueof-similar-to-keyof-in-typescript may be helpful.

Reply:
How about put the generic in the class, like this:

```
export class TableColumnFactory<T extends object, K extends keyof T> {
  transPrefix: string
  bar: BarColumn
  rows?: T[]

  constructor(transKeyPrefix: string, rows?: T[]) {
    this.transPrefix = transKeyPrefix
    this.rows = rows
    this.bar = new BarColumn(this)
  }
```

Not sure whether work, I am trying it.

Reply:
Done as @breeswish 's solution.